src/KOKKOS/compute_reaxff_bonds_kokkos.cpp

```diff
@@ -0,0 +1,192 @@
+// clang-format off
+/* ----------------------------------------------------------------------
+   LAMMPS - Large-scale Atomic/Molecular Massively Parallel Simulator
+   https://www.lammps.org/, Sandia National Laboratories
+   LAMMPS development team: developers@lammps.org
+
+   Copyright (2003) Sandia Corporation.  Under the terms of Contract
+   DE-AC04-94AL85000 with Sandia Corporation, the U.S. Government retains
+   certain rights in this software.  This software is distributed under
+   the GNU General Public License.
+
+   See the README file in the top-level LAMMPS directory.
+------------------------------------------------------------------------- */
+
+/* ----------------------------------------------------------------------
+   Contributing author: Richard Berger (LANL)
+------------------------------------------------------------------------- */
+
+#include "compute_reaxff_bonds_kokkos.h"
+#include "atom.h"
+#include "molecule.h"
+#include "update.h"
+#include "force.h"
+#include "memory.h"
+#include "error.h"
+#include "neigh_list.h"
+
+#include "memory_kokkos.h"
+#include "pair_reaxff_kokkos.h"
+#include "reaxff_api.h"
+
+using namespace LAMMPS_NS;
+using namespace ReaxFF;
+
+/* ---------------------------------------------------------------------- */
+
+template<class DeviceType>
+ComputeReaxFFBondsKokkos<DeviceType>::ComputeReaxFFBondsKokkos(LAMMPS *lmp, int narg, char **arg) :
+  ComputeReaxFFBonds(lmp, narg, arg),
+  nbuf(-1), buf(nullptr)
+{
+}
+
+
+/* ---------------------------------------------------------------------- */
+
+template<class DeviceType>
+ComputeReaxFFBondsKokkos<DeviceType>::~ComputeReaxFFBondsKokkos()
+{
+  memoryKK->destroy_kokkos(k_buf, buf);
+}
+
+/* ---------------------------------------------------------------------- */
+
+template<class DeviceType>
+void ComputeReaxFFBondsKokkos<DeviceType>::init()
+{
+  reaxff = dynamic_cast<PairReaxFF*>(force->pair_match("^reax../kk",0));
+  if (reaxff == nullptr) error->all(FLERR,"Cannot use compute reaxff/bonds without "
+                  "pair_style reaxff/kk");
+}
+
+/* ---------------------------------------------------------------------- */
+template<class DeviceType>
+void ComputeReaxFFBondsKokkos<DeviceType>::compute_bonds()
+{
+  if (atom->nlocal > nlocal) {
+    memory->destroy(array_atom);
+    nlocal = atom->nlocal;
+    memory->create(array_atom, nlocal, 7, "reaxff/bonds:array_atom");
+  }
+
+  // retrieve bond information from kokkos pair style. the data potentially
+  // lives on device. it is copied into buf on the host in a condensed format
+  // compute_local and compute_atom then expand the data from this buffer into
+  // appropiate arrays for consumption by others (e.g. dump local, dump custom
+  // or library interface)
+  int maxnumbonds = 0;
+  if (reaxff->execution_space == Device)
+    device_pair()->FindBond(maxnumbonds);
```

Comment:
Dynamic cast is pretty expensive, I would move this up to init() and store the result.

Reply:
Yes, but this is two calls during `compute_bonds`. The main purpose of the helpers is to to hide the type via auto. I doubt it's more expensive then calling kernels.